### ISSUE TITLE
Mishandling of wildcard in query_string queries

### ISSUE THREAD

**Issue by esatterwhite** [quickwit-oss/quickwit#3918]
**Describe the bug**
When using the `query_string` query in the elasticsearch compatible DSL, it seems the handling of trailing wild cards is different, or missing

Given the query
```json
        {
          "query_string": {
            "fields": [
              "app"
            ],
            "query": "plushies*",
            "lenient": true
          }
        }
```

Elasticsearch will return documents where the value of `app` is specifically `plushies` as it matches the `plushies*` query. Quick wit does not. I've tried the wildcard in several character positions, and only when omitted entirely will the expected results return

**Steps to reproduce (if applicable)**

1. insert a document when a field `app` with value `plushies`
2. attempt to match the document using a `query_string` with a query of `plushies*`

**Expected behavior**
Wild cards should be supported and match all documents that include the root value preceding the wildcard.

**Index Settings**
```json
{
  "field_mappings": [
    {
      "name": "_account",
      "type": "text",
      "fast": {
        "normalizer": "raw"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_app",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_bid",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_cluster",
      "type": "text",
      "fast": {
        "normalizer": "raw"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_env",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_file",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_hash",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_host",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_ingester",
      "type": "text",
      "fast": {
        "normalizer": "raw"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_ip",
      "type": "ip",
      "fast": true,
      "indexed": true,
      "stored": true
    },
    {
      "name": "_ipremote",
      "type": "ip",
      "fast": true,
      "indexed": true,
      "stored": true
    },
    {
      "name": "_label",
      "type": "json",
      "expand_dots": true,
      "fast": {
        "normalizer": "raw"
      },
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_lid",
      "type": "u64",
      "coerce": true,
      "fast": true,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "_line",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "source_code_default"
    },
    {
      "name": "_logtype",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_mac",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_meta",
      "type": "json",
      "expand_dots": true,
      "fast": {
        "normalizer": "raw"
      },
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_mezmo_line_size",
      "type": "u64",
      "coerce": true,
      "fast": true,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "_multiline",
      "type": "bool",
      "fast": true,
      "indexed": true,
      "stored": true
    },
    {
      "name": "_parsefail",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_parserids",
      "type": "array<text>",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_rawline",
      "type": "array<text>",
      "fast": false,
      "fieldnorms": false,
      "indexed": false,
      "stored": true
    },
    {
      "name": "_tag",
      "type": "array<text>",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "_ts",
      "type": "datetime",
      "fast": true,
      "indexed": true,
      "input_formats": [
        "unix_timestamp"
      ],
      "output_format": "unix_timestamp_millis",
      "precision": "milliseconds",
      "stored": true
    },
    {
      "name": "action",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "agent",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "at",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "auth",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "bytes",
      "type": "u64",
      "coerce": true,
      "fast": true,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "clientip",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "connect",
      "type": "u64",
      "coerce": true,
      "fast": true,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "container",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "containerid",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "dyno",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "facility",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "fwd",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "hostname",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "httpversion",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "logsource",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "message",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "meta",
      "type": "json",
      "expand_dots": true,
      "fast": {
        "normalizer": "raw"
      },
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "method",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "namespace",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "node",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "path",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "pid",
      "type": "u64",
      "coerce": true,
      "fast": false,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "pod",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "program",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "referrer",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "request",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "response",
      "type": "u64",
      "coerce": true,
      "fast": true,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "service",
      "type": "u64",
      "coerce": true,
      "fast": true,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "space",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "status",
      "type": "u64",
      "coerce": true,
      "fast": true,
      "indexed": true,
      "output_format": "string",
      "stored": true
    },
    {
      "name": "timestamp",
      "type": "text",
      "fast": {
        "normalizer": "raw"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "user",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "verb",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    {
      "name": "k8sclusterkey",
      "type": "text",
      "fast": {
        "normalizer": "lowercase"
      },
      "fieldnorms": false,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    }
  ],
  "tag_fields": [],
  "store_source": false,
  "index_field_presence": false,
  "timestamp_field": "_ts",
  "mode": "dynamic",
  "dynamic_mapping": {
    "indexed": true,
    "tokenizer": "raw",
    "record": "basic",
    "stored": true,
    "expand_dots": true,
    "fast": {
      "normalizer": "raw"
    }
  },
  "max_num_partitions": 200,
  "tokenizers": []
}
```

**Comment by fulmicoton** [comment]
The `*` can happen in the middle of a token. We need to identify the prefix to restrict the IO that we have to do for this, and then use the automaton mechanic.

We might as well start supporting regex queries, and extract prefixes (probably from the regular expression, as opposed to from the regex.)

**Comment by esatterwhite** [comment]
@fulmicoton Any traction on this one. This has been coming up a lot

**Comment by fulmicoton** [comment]
@trinity-1686a can you have a look?

**Comment by trinity-1686a** [comment]
after discussing, the issue at the time where the issue was opened was that we didn't support wildcards at all. It has since been partially implemented in #4059, but still misses features.
The feature which caused the UP is that wildcard queries don't lowercase as they should. I'll open a PR for that.
There is at least one other issue with wildcards which is we don't support wildcard anywhere but at the last position (`pl*shies` isn't okay). That last part will probably be implemented by leveraging regex support when it appears

**Comment by trinity-1686a** [comment]
the issue from this ticket is fixed, i've opened tickets for other issues one might have with our current support of wildcards #5306 #5307
